File: 0164-Maximum Gap/0164-Maximum Gap.cpp
class Solution {
public:
    int maximumGap(vector<int>& nums) {
        int n = nums.size();
        if (n < 2) {
            return 0;
        }
        
        int maxI = *max_element(nums.begin(), nums.end());
        int minI = *min_element(nums.begin(), nums.end());
        
        int gap = (maxI - minI - 1) / (n - 1) + 1;
        vector<int> bucketMin(n - 1, INT_MAX);
        vector<int> bucketMax(n - 1, INT_MIN);
        
        int maxGap = 0;
        for (int i : nums) {
            if (i == minI || i == maxI) {
                continue;
            }
            
            int index = (i - minI) / gap;
            bucketMin[index] = min(bucketMin[index], i);
            bucketMax[index] = max(bucketMax[index], i);
        }
        
        int previous = minI;
        for (int i = 0; i < n - 1; ++i) {
            if (bucketMin[i] == INT_MAX) {
                continue;
            }
            
            maxGap = max(maxGap, bucketMin[i] - previous);
            previous = bucketMax[i];
        }
        
        return max(maxGap, maxI - previous);
    }
};

File: 0101-0200/0164-Maximum Gap/0164-Maximum Gap.py
import math
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        if len(nums) < 2:
            return 0

        maxNum = max(nums)
        minNum = min(nums)
        if maxNum - minNum <= 1:
            return maxNum - minNum
        gap = math.ceil((maxNum - minNum + 1) / (len(nums) - 1))
        bucketMax = [-math.inf] * (len(nums) - 1)
        bucketMin = [math.inf] * (len(nums) - 1)
        for num in nums:
            index = (num - minNum) // gap
            bucketMax[index] = max(bucketMax[index], num)
            bucketMin[index] = min(bucketMin[index], num)

        maxGap = 0
        previous = minNum
        for i in range(len(bucketMax)):
            if bucketMin[i] == math.inf:
                continue
            maxGap = max(maxGap, bucketMin[i] - previous)
            previous = bucketMax[i]

        return max(maxGap, bucketMax[-1] - bucketMin[-1])
